File: android/app/src/main/java/com/turbometa/rayban/managers/QuickVisionModeManager.kt
package com.turbometa.rayban.managers

import android.content.Context
import android.content.SharedPreferences
import com.turbometa.rayban.R
import com.turbometa.rayban.models.QuickVisionMode
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Quick Vision Mode Manager
 * 快速识图模式管理器 - 管理当前模式、自定义提示词、翻译目标语言
 */
class QuickVisionModeManager private constructor(private val context: Context) {

    companion object {
        private const val PREFS_NAME = "quick_vision_prefs"
        private const val KEY_MODE = "quick_vision_mode"
        private const val KEY_CUSTOM_PROMPT = "quick_vision_custom_prompt"
        private const val KEY_TRANSLATE_TARGET_LANGUAGE = "quick_vision_translate_target_language"

        @Volatile
        private var instance: QuickVisionModeManager? = null

        fun getInstance(context: Context): QuickVisionModeManager {
            return instance ?: synchronized(this) {
                instance ?: QuickVisionModeManager(context.applicationContext).also { instance = it }
            }
        }

        // 支持的翻译目标语言
        val supportedLanguages: List<Pair<String, String>> = listOf(
            "zh-CN" to "中文",
            "en-US" to "English",
            "ja-JP" to "日本語",
            "ko-KR" to "한국어",
            "fr-FR" to "Français",
            "de-DE" to "Deutsch",
            "es-ES" to "Español",
            "it-IT" to "Italiano",
            "pt-BR" to "Português",
            "ru-RU" to "Русский"
        )
    }

    private val prefs: SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    // 当前模式
    private val _currentMode = MutableStateFlow(loadMode())
    val currentMode: StateFlow<QuickVisionMode> = _currentMode.asStateFlow()

    // 自定义提示词
    private val _customPrompt = MutableStateFlow(loadCustomPrompt())
    val customPrompt: StateFlow<String> = _customPrompt.asStateFlow()

    // 翻译目标语言
    private val _translateTargetLanguage = MutableStateFlow(loadTranslateTargetLanguage())
    val translateTargetLanguage: StateFlow<String> = _translateTargetLanguage.asStateFlow()

    private fun loadMode(): QuickVisionMode {
        val modeId = prefs.getString(KEY_MODE, QuickVisionMode.STANDARD.id) ?: QuickVisionMode.STANDARD.id
        return QuickVisionMode.fromId(modeId)
    }

    private fun loadCustomPrompt(): String {
        return prefs.getString(KEY_CUSTOM_PROMPT, context.getString(R.string.quickvision_custom_default)) ?: ""
    }

    private fun loadTranslateTargetLanguage(): String {
        return prefs.getString(KEY_TRANSLATE_TARGET_LANGUAGE, "zh-CN") ?: "zh-CN"
    }

    fun setMode(mode: QuickVisionMode) {
        _currentMode.value = mode
        prefs.edit().putString(KEY_MODE, mode.id).apply()
    }

    fun setCustomPrompt(prompt: String) {
        _customPrompt.value = prompt
        prefs.edit().putString(KEY_CUSTOM_PROMPT, prompt).apply()
    }

    fun setTranslateTargetLanguage(languageCode: String) {
        _translateTargetLanguage.value = languageCode
        prefs.edit().putString(KEY_TRANSLATE_TARGET_LANGUAGE, languageCode).apply()
    }

    /**
     * 获取当前模式的完整提示词
     */
    fun getPrompt(): String {
        return when (_currentMode.value) {
            QuickVisionMode.CUSTOM -> _customPrompt.value
            QuickVisionMode.TRANSLATE -> getTranslatePrompt()
            else -> _currentMode.value.getPrompt(context)
        }
    }

    /**
     * 获取指定模式的提示词
     */
    fun getPrompt(mode: QuickVisionMode): String {
        return when (mode) {
            QuickVisionMode.CUSTOM -> _customPrompt.value
            QuickVisionMode.TRANSLATE -> getTranslatePrompt()
            else -> mode.getPrompt(context)
        }
    }

    /**
     * 获取翻译模式的提示词（包含目标语言）
     */
    private fun getTranslatePrompt(): String {
        val targetLanguageName = supportedLanguages.find { it.first == _translateTargetLanguage.value }?.second ?: "中文"
        val basePrompt = context.getString(R.string.prompt_quickvision_translate)
        return basePrompt.replace("{LANGUAGE}", targetLanguageName)
    }

    /**
     * 获取目标语言名称
     */
    fun getTranslateTargetLanguageName(): String {
        return supportedLanguages.find { it.first == _translateTargetLanguage.value }?.second ?: "中文"
    }
}

File: CameraAccess/Managers/QuickVisionModeManager.swift
/*
 * Quick Vision Mode Manager
 * 快速识图模式管理器 - 管理当前模式、自定义提示词、翻译目标语言
 */

import Foundation
import SwiftUI

class QuickVisionModeManager: ObservableObject {
    static let shared = QuickVisionModeManager()

    private let userDefaults = UserDefaults.standard
    private let modeKey = "quickVisionMode"
    private let customPromptKey = "quickVisionCustomPrompt"
    private let translateTargetLanguageKey = "quickVisionTranslateTargetLanguage"

    @Published var currentMode: QuickVisionMode {
        didSet {
            userDefaults.set(currentMode.rawValue, forKey: modeKey)
            print("📋 [QuickVisionModeManager] 模式已切换: \(currentMode.displayName)")
        }
    }

    @Published var customPrompt: String {
        didSet {
            userDefaults.set(customPrompt, forKey: customPromptKey)
        }
    }

    @Published var translateTargetLanguage: String {
        didSet {
            userDefaults.set(translateTargetLanguage, forKey: translateTargetLanguageKey)
        }
    }

    // 支持的翻译目标语言
    static let supportedLanguages: [(code: String, name: String)] = [
        ("zh-CN", "中文"),
        ("en-US", "English"),
        ("ja-JP", "日本語"),
        ("ko-KR", "한국어"),
        ("fr-FR", "Français"),
        ("de-DE", "Deutsch"),
        ("es-ES", "Español"),
        ("it-IT", "Italiano"),
        ("pt-BR", "Português"),
        ("ru-RU", "Русский")
    ]

    private init() {
        // 加载保存的模式
        if let savedMode = userDefaults.string(forKey: modeKey),
           let mode = QuickVisionMode(rawValue: savedMode) {
            self.currentMode = mode
        } else {
            self.currentMode = .standard
        }

        // 加载自定义提示词
        self.customPrompt = userDefaults.string(forKey: customPromptKey) ?? "quickvision.custom.default".localized

        // 加载翻译目标语言（默认跟随系统语言）
        if let savedLanguage = userDefaults.string(forKey: translateTargetLanguageKey) {
            self.translateTargetLanguage = savedLanguage
        } else {
            self.translateTargetLanguage = LanguageManager.staticApiLanguageCode
        }
    }

    // MARK: - Get Current Prompt

    /// 获取当前模式的完整提示词
    func getPrompt() -> String {
        switch currentMode {
        case .custom:
            return customPrompt
        case .translate:
            return getTranslatePrompt()
        default:
            return currentMode.prompt
        }
    }

    /// 获取指定模式的提示词
    func getPrompt(for mode: QuickVisionMode) -> String {
        switch mode {
        case .custom:
            return customPrompt
        case .translate:
            return getTranslatePrompt()
        default:
            return mode.prompt
        }
    }

    /// 获取翻译模式的提示词（包含目标语言）
    private func getTranslatePrompt() -> String {
        let targetLanguageName = Self.supportedLanguages.first { $0.code == translateTargetLanguage }?.name ?? "中文"
        let basePrompt = "prompt.quickvision.translate".localized
        return basePrompt.replacingOccurrences(of: "{LANGUAGE}", with: targetLanguageName)
    }

    // MARK: - Mode Management

    func setMode(_ mode: QuickVisionMode) {
        currentMode = mode
    }

    func setCustomPrompt(_ prompt: String) {
        customPrompt = prompt
    }

    func setTranslateTargetLanguage(_ languageCode: String) {
        translateTargetLanguage = languageCode
    }

    // MARK: - Static Access (for non-SwiftUI contexts)

    static var staticCurrentMode: QuickVisionMode {
        return shared.currentMode
    }

    static var staticPrompt: String {
        return shared.getPrompt()
    }
}
